### DISCUSSION TITLE
RPi 3b & zero 2 kernel bug: swapped colors ⚠️

### DISCUSSION THREAD

**Discussion by awawa-dev** [awawa-dev/HyperHDR#848] [Q&A] [answered]
When using a USB grabber color swapping may occur (blue with red, red with green...).

correct one:
![obraz](https://github.com/awawa-dev/HyperHDR/assets/69086569/49c88410-6d45-4a2f-ba4a-4608aff7e2bf)

captured image:
![obraz](https://github.com/awawa-dev/HyperHDR/assets/69086569/d1bc7c48-6a4b-44b4-b66d-14a67e7d603f)

Symptoms described by our users:
https://github.com/awawa-dev/HyperHDR/discussions/867
https://github.com/awawa-dev/HyperHDR/discussions/844
https://github.com/awawa-dev/HyperHDR/discussions/834
https://github.com/awawa-dev/HyperHDR/discussions/797

Affected devices:
USB grabbers with Raspberry Pi OS Bookworm, 6.6.x Kernel
RPi3b and probably similar A53 boards, on the other hand RPi4 is not affected

⚠️ Not sure if the RPi team is aware of the problem or is planning to fix it. So please report the issue to them to fix it, or stay on Kernel 6.1 or buy RPi4/5 :shrug: ⚠️ Or buy Intel N100 mini PC :smiling_imp:

Known solutions:
- downgrade the Kernel to 6.1.x
- use our Debian Bullseye based official HyperHDR SD images (https://github.com/awawa-dev/HyperHDR/wiki/Installation#raspberrypi-os-sd-card-image-installation)
- use older Raspberry Pi OS Bookworm release ex. aarch64: https://downloads.raspberrypi.org/raspios_lite_arm64/images/raspios_lite_arm64-2023-12-11/ or armhf: [https://downloads.raspberrypi.org/raspios_lite_armhf/images/raspios_lite_armhf-2023-12-11/](https://downloads.raspberrypi.org/raspios_lite_armhf/images/raspios_lite_armhf-2023-12-11/)  or even older version from https://downloads.raspberrypi.org server (after install do not upgrade the kernel!)

**Comment by SodaPop-droid** [comment]
New to this sort of thing - How would I go about downgrading the kernel?
Is it a type of code i could copy across into PuTTY?

**Reply by awawa-dev** [reply]
https://github.com/awawa-dev/HyperHDR/discussions/867#discussioncomment-9561187

**Reply by SodaPop-droid** [reply]
I thought this one would be referenced, I'm uncertain how the Hash works.
Do i copy the blue hyperlink across or go into it and source part of the code from somewhere?

**Reply by awawa-dev** [reply]
There you have the sudo rpi-update command with a hash to execute. As I noted, this has not been tested by me and I am only relying on the general RPi manual so you can read it too before proceeding.

**Reply by SodaPop-droid** [reply]
Aa I see, thank you for your help

**Comment by zigomatichub** [comment]
Trying as I'm facing the issue after upgrade the firmware on my raspberry pi zero 2W + update HyperHDR to 20.0.0. 
All was done together,then difficult for me to see where is the issue.
 
I found commits on https://github.com/raspberrypi/rpi-firmware/commits/master/?after=1ec318fceecbb3d518dc889a3ed8369c5e1aac37+34

And put back with 6.1.77
sudo rpi-update 5fc4f643d2e9c5aa972828705a902d184527ae3f

Now in SDR, It's back OK. 
But HDR, still color is not aligned.
I think in the HyperHDR 19, it was also the case where I never get that corrected. HDR it's link to kernel as it was also before...
It's may link to my splitter ROFAVEZCO HDMI 2.0 Splitter 4k 60hz 1080p 120hz.
I'm surprise as well that a kernel can give trouble like this and I don't know how awawa-dev found that relation but it's a good one 👍

**Reply by LEDExperiments** [reply]
Thank you so much for posting a concrete command what to do to fix this bug! I just applied _ sudo rpi-update 5fc4f643d2e9c5aa972828705a902d184527ae3f_ and now I get the right colors 👍

**Comment by sysTim1977** [comment]
One of the latest funktionable Kernel is 6.1.74

root@yourhost# rpi-update d86b5843d68b9972a5430a6d3da1b271cfc83521

**Comment by sysTim1977** [comment]
Solved!!!
but i have no idea how to adapt, please help me.

My solution was to set the "User Controls" of v4l2 manualy

user@system:$  v4l2-ctl --set-ctrl=brightness=50 --set-ctrl=contrast=50 --set-ctrl=saturation=50 --set-ctrl=hue=50

Explanation:
I compared the output of "v4l2-ctl --all" on (working) kernel 6.1 with the output on the misworking kernel 6.6

---begin output---
kernel-6.1.x:
...
User Controls
brightness 0x00980900 (int)    : min=0 max=100 step=1 default=50 value=50
contrast 0x00980901 (int)    : min=0 max=100 step=1 default=50 value=50
saturation 0x00980902 (int)    : min=0 max=100 step=1 default=50 value=50
hue 0x00980903 (int)    : min=0 max=100 step=1 default=50 value=50

kernel-6.6.x:
...
User Controls
brightness 0x00980900 (int)    : min=0 max=100 step=1 default=-8193 value=0
contrast 0x00980901 (int)    : min=0 max=100 step=1 default=57343 value=100
saturation 0x00980902 (int)    : min=0 max=100 step=1 default=57343 value=100
hue 0x00980903 (int)    : min=0 max=100 step=1 default=-8193 value=0
---end output---

So I tried to setup the "User-Controls" Manually with: "v4l2-ctl --set-ctrl=brightness=50 --set-ctrl=contrast=50 --set-ctrl=saturation=50 --set-ctrl=hue=50"; like Kernel 6.1.x

It works! The collours are not swapped any longer.

How can I adapt this on boot?
(RaspberryPiOS-Lite on PiZero2W, Kernel 6.6.56-v8+)
Thanks

**Reply by awawa-dev** [reply]
You can set these values to 50 in HyperHDR and test it. A pretty nasty kernel driver bug... Have you reported this issue to the RPi team?

**Reply by sysTim1977** [reply]
No I have not reported this.
The manual setting in the webOS works fine.
Thanks. 
Perhaps you can add this hint to the Discussion

**Reply by awawa-dev** [reply]
In the grabber settings: hardware brightness, contrast, saturation, hue...But this workaround will work only for ms2130 if not fixed in the kernel. Other grabbers usually have different default values.

**Reply by sysTim1977** [reply]
How can I report this issue to the RPi Team?

**Reply by awawa-dev** [reply]
You can try here: https://github.com/raspberrypi/linux/issues Thank you for your investigation 👍

**Comment by admin** [comment, marked as answer]
### Update on USB Grabber Color Bug (RPi 3 / Zero 2W)

The swapped/inverted colors issue with USB grabbers on kernel 6.6+ is still present. There's an open ticket here: https://github.com/raspberrypi/linux/issues/7410

Interestingly, despite the problem being reported over 1.5 years ago (including on the official RPi forum with the same engineer involved there), and even with provided logs from various grabbers showing broken UVC control defaults, the discussion is now focused on "it works for us" and suggestions to use a 32-bit kernel or upgrade the board (and trash your RPi e-waste into the bin or use it for a project that doesn't require a stable USB controller).

I am no longer proxying these reports as I don't use the affected devices. If you're affected, it's in your best interest to report it in the linked ticket, as it's the probably last chance for this issue to be fixed in the mainline RPi kernel.
